File: deepspeed/inference/v2/kernels/core_ops/bias_activations/bias_activation.py
# Copyright (c) Microsoft Corporation.
# SPDX-License-Identifier: Apache-2.0

# DeepSpeed Team

from typing import Optional

import torch

from ....inference_utils import ActivationType, DtypeEnum
from deepspeed.ops.op_builder import InferenceCoreBuilder
from ... import DSKernelBase


class CUDABiasActivation(DSKernelBase):
    """
    CUDA implementation of bias activation kernel. This kernel should be deprecated once
    we are fusing the bias activation into the linear kernel in all scenarios.
    """

    supported_dtypes = [DtypeEnum.fp16, DtypeEnum.bf16]
    supported_act_fns = [ActivationType.IDENTITY, ActivationType.GELU, ActivationType.RELU, ActivationType.SILU]

    def __init__(self, channels: int, dtype: DtypeEnum, act_fn: ActivationType) -> None:
        """
        Compile and validate for the fused bias-activation kernel.

        Parameters:
            channels (int): Number of channels to expect in the activation.
            dtype (torch.dtype): Data type for the input/output. Supported values
                are DtypeEnum.fp16 and DtypeEnum.bf16.
            act_fn (ActivationType): Activation function to use. Only IDENTITY, GELU, RELU, and SILU are supported.
        """

        if channels % 8 != 0:
            raise ValueError("channels must be divisible by 8")

        if DtypeEnum(dtype) not in CUDABiasActivation.supported_dtypes:
            raise ValueError("Unsupported data type: {}, supported_dtypes are {}".format(
                dtype, CUDABiasActivation.supported_dtypes))

        act_fn = ActivationType(act_fn)
        if act_fn not in CUDABiasActivation.supported_act_fns:
            raise ValueError("Unsupported activation function: {}, supported_act_fns are {}".format(
                act_fn, CUDABiasActivation.supported_act_fns))

        inf_module = InferenceCoreBuilder().load()
        self.kernel = inf_module.bias_activation
        self.act_fn = act_fn

    def __call__(self, activation: torch.Tensor, bias: Optional[torch.Tensor] = None) -> torch.Tensor:
        """
        Add an optional bias and perform the non-linear activation function.

        Parameters:
            activation (torch.Tensor): Input tensor of shape [tokens, channels]
            bias (torch.Tensor): Optional bias tensor of shape [channels]

        Returns:
            activation that has been updated in-place
        """
        self.kernel(activation, bias, self.act_fn.value)

File: deepspeed/inference/v2/kernels/core_ops/bias_activations/bias_activation.cpp
// Copyright (c) Microsoft Corporation.
// SPDX-License-Identifier: Apache-2.0

// DeepSpeed Team

#include "bias_activation.h"
#include <c10/cuda/CUDAStream.h>
#include "ds_kernel_utils.h"

#ifdef BF16_AVAILABLE
#define DTYPE_SWITCH(DTYPE, ...)                                        \
    [&] {                                                               \
        if (DTYPE == torch::kFloat16) {                                 \
            using scalar_t = __half;                                    \
            return __VA_ARGS__();                                       \
        } else if (DTYPE == torch::kBFloat16) {                         \
            using scalar_t = __nv_bfloat16;                             \
            return __VA_ARGS__();                                       \
        } else {                                                        \
            TORCH_CHECK(false, "Unsupported dtype for BiasActivation"); \
        }                                                               \
    }()
#else
#define DTYPE_SWITCH(DTYPE, ...)                                        \
    [&] {                                                               \
        if (DTYPE == torch::kFloat16) {                                 \
            using scalar_t = __half;                                    \
            return __VA_ARGS__();                                       \
        } else {                                                        \
            TORCH_CHECK(false, "Unsupported dtype for BiasActivation"); \
        }                                                               \
    }()
#endif

/*
In-place bias and activation fusion kernel.
*/
void bias_activation(torch::Tensor& activation,
                     c10::optional<torch::Tensor>& bias,
                     const int32_t act_type)
{
    const ActivationType atype = static_cast<ActivationType>(act_type);
    const int32_t rows = activation.size(0);
    const int32_t cols = activation.size(1);

    TORCH_CHECK(atype == ActivationType::GELU || atype == ActivationType::RELU ||
                    atype == ActivationType::SILU || atype == ActivationType::IDENTITY,
                "Unsupported activation type for BiasActivation");
    TORCH_CHECK(activation.dim() == 2, "BiasActivation only supports 2D activation tensors");

    DTYPE_SWITCH(activation.scalar_type(), [&] {
        scalar_t* activation_ptr = reinterpret_cast<scalar_t*>(activation.data_ptr());

        const scalar_t* bias_ptr;
        if (bias.has_value()) {
            TORCH_CHECK(activation.scalar_type() == bias.value().scalar_type(),
                        "BiasActivation activation and bias must have same dtype");
            bias_ptr = reinterpret_cast<const scalar_t*>(bias.value().data_ptr());
        } else {
            bias_ptr = nullptr;
        }

        if (atype == ActivationType::IDENTITY && bias_ptr == nullptr) { return; }

        launch_bias_activation<scalar_t>(
            activation_ptr, bias_ptr, rows, cols, atype, c10::cuda::getCurrentCUDAStream());
    });
}
